File: week01/ProductOfArrayExceptSelf.java
public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        int product = nums[n-1];
        
        result[0] = nums[0];
        for(int i=1; i<n; i++){
            result[i] = result[i-1] * nums[i];
        }
        
        
        result[n-1] = result[n-2];
        for(int i=n-2; i>0; i--){
            result[i] = result[i-1] * product;
            product*=nums[i];
        }
        result[0] = product;
        
        return result;
    }

File: week01/product_of_array_except_self.cpp
// TC: O(N)
// MC: O(N)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> prefixProducts(nums.size(), 1), suffixProducts(nums.size(), 1), products;
        for (int i = 1; i < nums.size(); i++)
        {
            prefixProducts[i] = prefixProducts[i - 1] * nums[i - 1];
        }
        for (int i = nums.size() - 2; i >= 0; i--)
        {
            suffixProducts[i] = suffixProducts[i + 1] * nums[i + 1];
        }
        for (int i = 0; i < nums.size(); i++)
        {
            products.push_back(prefixProducts[i] * suffixProducts[i]);
        }
        return products;
    }
};